src/messaging/tests/TestReliableMessageProtocol.cpp

```diff
@@ -29,24 +29,62 @@
 #include <messaging/ReliableMessageManager.h>
```

Comment:
I can't find any tests for dropped packets, can you highlight those?  Especially where another message needs to be sent, has been allocated (i.e. might have been re-used by the underlying allocator).

Reply:
Those kind of testings are part of integration test, we have added the following debug functions and FaultInjection mechanism to test those scenario:
 bool ShouldDropAckDebug() const;
 void SetDropAckDebug(bool inDropAckDebug);

Current unit testing framework are doing functional level test, we have pretty comprehensive testing for re-transmission and dropped packets scenario in Weave, but they are all in happy/cirque.

Reply:
dropped packets should be accomplished in unit tests with a mocking transport.  there's already a loopback, right?  introduce some facility for drops

Reply:
Added the test case "CheckResendMessage" to simulate dropping the sending packets, I am still not able to simulate dropping the received packets with loopback. This is because of how the exchange is designed.

When the test program first sends a packet to anther node, it first creates an exchange with peerNode of the Destination node ID 111222333. When the loopback transport send the packet back to it, the ExchangeManager will automatically create an new exchange with peerNode ID 123654 using the source Node ID of the packet. Now the new created exchange is invalid since its peerNode ID is same as its own NodeID. We have to run two independent programs with different Node ID to fully simulating the end-to-end communication.